File: libs/fleet/sdk-bindings/examples/kotlin/AppControlled.kt
import ai.cua.cyclops.sdk.*
import ai.cua.cyclops.sdk.schema.*
import kotlinx.coroutines.runBlocking
import java.util.Collections

private data class Expected(val method: String, val url: String, val headers: List<Pair<String, String>>, val body: ByteArray?, val status: UShort, val response: String)
private val generatedClaimNamePattern = Regex("claim-[a-z0-9](?:[-a-z0-9]*[a-z0-9])?")
private val claimNameFieldPattern = Regex("\"name\":\"(claim-[^\"]+)\"")

private fun normalizedGeneratedClaimBody(body: ByteArray?): ByteArray? {
    val text = body?.decodeToString() ?: return null
    val match = claimNameFieldPattern.find(text) ?: return null
    val name = match.groupValues[1]
    check(name.length <= 63 && generatedClaimNamePattern.matches(name))
    return text.replaceRange(match.groups[1]!!.range, "claim-generated").encodeToByteArray()
}

private class ScriptedHttpClient : HttpClient {
    private val lock = Any()
    private val expected = ArrayDeque(listOf(
        Expected("POST", "https://keycloak.invalid/token", listOf("accept" to "application/json", "content-type" to "application/x-www-form-urlencoded", "authorization" to "Basic Y2xpZW50LWlkOmNsaWVudC1zZWNyZXQ="), "grant_type=client_credentials".encodeToByteArray(), 200u, "{\"access_token\":\"offline-token\",\"expires_in\":3600}"),
        Expected("POST", "https://cyclops.invalid/api/namespaces", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"name\":\"default\"}".encodeToByteArray(), 201u, "{}"),
        Expected("POST", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxWarmPool\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"replicas\":1,\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":null}".encodeToByteArray(), 201u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxWarmPool\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"replicas\":1,\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":null}"),
        Expected("POST", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"claim-1\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"},\"bindDeadline\":900},\"status\":null}".encodeToByteArray(), 201u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":{\"phase\":\"Bound\",\"sandbox\":{\"name\":\"offline-sandbox\"}}}"),
        Expected("GET", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 200u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":{\"phase\":\"Bound\",\"sandbox\":{\"name\":\"offline-sandbox\"}}}"),
        Expected("GET", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxtemplates/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 200u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxTemplate\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"vmTemplate\":{\"containerDiskImage\":\"registry.example/desktop:offline\",\"services\":[{\"name\":\"mcp\",\"targetPort\":8080}]}}}"),
        Expected("POST", "https://cyclops.invalid/api/svc/default/offline-sandbox-mcp/mcp", listOf("X-Cua-Fleet-Claim" to "default", "authorization" to "Bearer offline-token"), "{\"offline\":true}".encodeToByteArray(), 202u, "offline service accepted"),
        Expected("DELETE", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, ""),
        Expected("DELETE", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, ""),
        Expected("DELETE", "https://cyclops.invalid/api/namespaces/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, "")
    ))
    override suspend fun execute(request: HttpRequest): HttpResponse = synchronized(lock) {
        val item = expected.removeFirst()
        check(request.method == item.method && request.url == item.url)
        check(request.headers.map { it.name to it.value } == item.headers)
        if (request.method == "POST" && request.url.endsWith("/osgymsandboxclaims")) {
            val actualBody = normalizedGeneratedClaimBody(request.body)
            val expectedBody = normalizedGeneratedClaimBody(item.body)
            check(actualBody != null && expectedBody != null && actualBody.contentEquals(expectedBody))
        } else {
            check((request.body == null && item.body == null) || (request.body != null && item.body != null && request.body.contentEquals(item.body)))
        }
        HttpResponse(item.status, emptyList(), item.response.encodeToByteArray())
    }
    fun assertExhausted() = check(expected.isEmpty())
}

fun main() = runBlocking {
    val spec = OsGymSandboxWarmPoolSpec(1u, SandboxTemplateRef("default"), null)
    val transport = ScriptedHttpClient()
    val credentials = CyclopsCredentials("client-id", "client-secret")
    val client = CyclopsClient.connect(CyclopsConfiguration("https://cyclops.invalid", "https://keycloak.invalid/token", credentials, 1uL, 1u, 1uL, 2u), transport)
    val pool = client.createPool(CreatePoolRequest("default", spec))
    val claim = client.createClaim(CreateClaimRequest(pool, ClaimSpec(SandboxTemplateRef(pool.metadata.name), null, null, null)))
    val sandbox = client.waitClaim(claim)
    val service = client.serviceRequest(sandbox, "mcp", "/mcp", HttpRequest("POST", "https://ignored.invalid/mcp", emptyList(), "{\"offline\":true}".encodeToByteArray(), null))
    client.deleteClaim(claim)
    client.deletePool(pool)
    check(pool.metadata.name == "default" && claim.metadata.name == "default" && sandbox.name == "offline-sandbox")
    check(service.status == 202.toUShort())
    transport.assertExhausted()
    println("pool=${pool.metadata.name} claim=${claim.metadata.name} sandbox=${sandbox.name} service_status=${service.status}")
}

File: libs/fleet/sdk-bindings/examples/ruby/app_controlled.rb
require 'json'
require 'cyclops_sdk'

require 'json'
require 'thread'
Expected = Struct.new(:method, :url, :headers, :body, :status, :response)
BASE = 'https://cyclops.invalid'; TOKEN = 'https://keycloak.invalid/token'; JSON_HEADERS = [['accept','application/json'],['content-type','application/json'],['authorization','Bearer offline-token']]
GENERATED_CLAIM_NAME = /\Aclaim-[a-z0-9](?:[-a-z0-9]*[a-z0-9])?\z/

def normalized_generated_claim_body(body)
  return nil if body.nil?
  value = JSON.parse(body)
  name = value.fetch('metadata').fetch('name')
  raise "invalid generated claim name: #{name.inspect}" unless name.bytesize <= 63 && GENERATED_CLAIM_NAME.match?(name)
  value['metadata']['name'] = 'claim-generated'
  JSON.generate(value).b
end

def pool_json; { apiVersion:'osgym.cua.ai/v1alpha1',kind:'OSGymSandboxWarmPool',metadata:{namespace:'default',name:'default',labels:nil},spec:{replicas:1,sandboxTemplateRef:{name:'default'}},status:nil }; end
def template_json; { apiVersion:'osgym.cua.ai/v1alpha1',kind:'OSGymSandboxTemplate',metadata:{namespace:'default',name:'default',labels:nil},spec:{vmTemplate:{containerDiskImage:'registry.example/desktop:offline',services:[{name:'mcp',targetPort:8080}]}} }; end
def claim_json(bound=false); value={apiVersion:'osgym.cua.ai/v1alpha1',kind:'OSGymSandboxClaim',metadata:{namespace:'default',name:'default',labels:nil},spec:{sandboxTemplateRef:{name:'default'}},status:nil}; value[:status]={phase:'Bound',sandbox:{name:'offline-sandbox'}} if bound; value; end
def queue
 pool_url="#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools/default"; template_url="#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxtemplates/default"; claim_url="#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default"; pool_body=JSON.generate(pool_json).b; claim_body='{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"claim-1","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"},"bindDeadline":900},"status":null}'.b
 [Expected.new('POST',TOKEN,[['accept','application/json'],['content-type','application/x-www-form-urlencoded'],['authorization','Basic Y2xpZW50LWlkOmNsaWVudC1zZWNyZXQ=']], 'grant_type=client_credentials'.b,200,{access_token:'offline-token',expires_in:3600}),Expected.new('POST',"#{BASE}/api/namespaces",JSON_HEADERS,'{"name":"default"}'.b,201,{}),Expected.new('POST',pool_url.sub(%r{/default\z}, ''),JSON_HEADERS,pool_body,201,pool_json),Expected.new('POST',claim_url.sub(%r{/default\z}, ''),JSON_HEADERS,claim_body,201,claim_json),Expected.new('GET',claim_url,JSON_HEADERS,nil,200,claim_json(true)),Expected.new('GET',template_url,JSON_HEADERS,nil,200,template_json),Expected.new('POST',"#{BASE}/api/svc/default/offline-sandbox-mcp/mcp",[['X-Cua-Fleet-Claim','default'],['authorization','Bearer offline-token']],'{"offline":true}'.b,202,'offline service accepted'),Expected.new('DELETE',claim_url,JSON_HEADERS,nil,204,''),Expected.new('DELETE',pool_url,JSON_HEADERS,nil,204,''),Expected.new('DELETE',"#{BASE}/api/namespaces/default",JSON_HEADERS,nil,204,'')]
end

class ScriptedHttpClient < FleetSdk::HttpClient
  def initialize; @expected=queue; @mutex=Mutex.new; end
  def execute(request)
    @mutex.synchronize do
      item=@expected.shift or raise 'unexpected request'
      actual=[request.method,request.url,request.headers.map{|h|[h.name,h.value]},request.body]
      expected=[item.method,item.url,item.headers,item.body]
      if request.method == 'POST' && request.url.end_with?('/osgymsandboxclaims')
        raise "request mismatch: #{actual.inspect}" unless actual.first(3) == expected.first(3) && normalized_generated_claim_body(request.body) == normalized_generated_claim_body(item.body)
      else
        raise "request mismatch: #{actual.inspect}" unless actual == expected
      end
      body=item.response.is_a?(String) ? item.response.b : JSON.generate(item.response).b
      FleetSdk::HttpResponse.new(status:item.status,headers:[],body:body)
    end
  end
  def assert_exhausted!; raise @expected.inspect unless @expected.empty?; end
end

spec = FleetSdk::OSGymSandboxWarmPoolSpec.new(replicas: 1, sandbox_template_ref: FleetSdk::SandboxTemplateRef.new(name: 'default'), autoscaling: nil)
transport = ScriptedHttpClient.new
credentials = FleetSdk::CyclopsCredentials.new('client-id', 'client-secret')
client = FleetSdk::CyclopsClient.connect(FleetSdk::CyclopsConfiguration.new(base_url: 'https://cyclops.invalid', token_url: 'https://keycloak.invalid/token', credentials: credentials, pool_poll_interval_ms: 1, pool_poll_limit: 1, claim_poll_interval_ms: 1, claim_poll_limit: 2), transport)
pool = client.create_pool(FleetSdk::CreatePoolRequest.new(namespace: 'default', spec: spec))
claim = client.create_claim(FleetSdk::CreateClaimRequest.new(pool: pool, spec: FleetSdk::ClaimSpec.new(sandbox_template_ref: FleetSdk::SandboxTemplateRef.new(name: pool.metadata.name), warmpool: nil, bind_deadline: nil, lifecycle: nil)))
sandbox = client.wait_claim(claim)
service = client.service_request(sandbox, 'mcp', '/mcp', FleetSdk::HttpRequestBuilder.new.method('POST').url('https://ignored.invalid/mcp').headers([]).body('{"offline":true}'.b).build)
client.delete_claim(claim)
client.delete_pool(pool)
transport.assert_exhausted!
raise 'unexpected service response' unless service.status == 202
puts "pool=#{pool.metadata.name} claim=#{claim.metadata.name} sandbox=#{sandbox.name} service_status=#{service.status}"
